### PR TITLE
fix: [SPG-2643]: adds a new path type to the gcs filter

### PR DESCRIPTION
**Pull request by rafaelmart1ns**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/46365)

### CODE REVIEW

**Comment by rafaelmart1ns** [comment]
trigger manager

File: 400-rest/src/test/java/software/wings/service/impl/BuildSourceServiceTest.java [hunk 1]
```diff
@@ -1446,14 +1446,34 @@ public void shouldListBuildsByArtifactStreamAndFilterPathWithSpecificPath() {
...
     when(artifactService.listArtifactsByArtifactStreamId(ACCOUNT_ID, ARTIFACT_STREAM_ID)).thenReturn(List.of(gcsArt));
     List<BuildDetails> builds =
         buildSourceService.listArtifactByArtifactStreamAndFilterPath(artifactList, artifactStream);
     assertThat(builds).hasSize(1);
   }
 
+  @Test
+  @Owner(developers = RAFAEL)
+  @Category(UnitTests.class)
+  public void shouldListBuildsByArtifactStreamAndFilterPathWithSpecificTypeFile() {
```
Comment (Fernando-Dourado) [inline 400-rest/src/test/java/software/wings/service/impl/BuildSourceServiceTest.java:1460]: Add one more test case when using multiple `*` into the path.

**Review by Fernando-Dourado — CHANGES_REQUESTED** [review]

File: 400-rest/src/main/java/software/wings/service/impl/BuildSourceServiceImpl.java [hunk 2]
```diff
@@ -924,8 +924,11 @@ public List<BuildDetails> listArtifactByArtifactStreamAndFilterPath(
         .filter(artifact -> {
           List<String> paths = ((GcsArtifactStream) artifactStream).getArtifactPaths();
           return paths.stream().anyMatch(path -> {
-            path = path.replace("*", "");
-            return artifact.getArtifactPath().startsWith(path);
+            String[] parts = path.split("\\*");
+            if (parts.length > 1) {
+              return artifact.getArtifactPath().startsWith(parts[0]) && artifact.getArtifactPath().endsWith(parts[1]);
+            }
+            return artifact.getArtifactPath().startsWith(parts[0]);
```
[the remark this answers is not in the document]
Comment (rafaelmart1ns) [reply 400-rest/src/main/java/software/wings/service/impl/BuildSourceServiceImpl.java:931]: this returns original path.

File: 400-rest/src/main/java/software/wings/service/impl/BuildSourceServiceImpl.java [hunk 3]
```diff
@@ -924,8 +924,11 @@ public List<BuildDetails> listArtifactByArtifactStreamAndFilterPath(
         .filter(artifact -> {
           List<String> paths = ((GcsArtifactStream) artifactStream).getArtifactPaths();
           return paths.stream().anyMatch(path -> {
-            path = path.replace("*", "");
-            return artifact.getArtifactPath().startsWith(path);
+            String[] parts = path.split("\\*");
+            if (parts.length > 1) {
```
[the remark this answers is not in the document]
Comment (rafaelmart1ns) [reply 400-rest/src/main/java/software/wings/service/impl/BuildSourceServiceImpl.java:928]: I apply the startwith when it had more than one *.

File: 400-rest/src/test/java/software/wings/service/impl/BuildSourceServiceTest.java (hunk 1 above)
Comment (rafaelmart1ns) [reply 400-rest/src/test/java/software/wings/service/impl/BuildSourceServiceTest.java:1460]: I will add

**Comment by rafaelmart1ns** [comment]
trigger manager

**Review by Fernando-Dourado — APPROVED** [review]

**Review by lasdox — APPROVED** [review]

**Pull request merged**